miniapps/nurbs/nurbs_naca_cmesh.cpp

```diff
@@ -0,0 +1,800 @@
+   delete kv_o2;
+   delete xyf[0];
+   delete xyf[1];
+
+   cout << endl << "Boundary identifiers:" << endl;
+   cout << "   1   Bottom" << endl;
+   cout << "   2   Inflow" << endl;
+   cout << "   3   Top" << endl;
+   cout << "   4   Outflow" << endl;
+   cout << "   5   Foil section" << endl;
+   cout << "=========================================================="<< endl;
+   cout << "  "<< mdim <<"D mesh generated: " <<mesh_file.c_str()<< endl ;
+   cout << "=========================================================="<< endl;
+
+   // Print mesh info to screen
+   cout << "=========================================================="<< endl;
+   cout << " Attempting to read mesh: " <<mesh_file.c_str()<< endl ;
+   cout << "=========================================================="<< endl;
+   Mesh *mesh = new Mesh(mesh_file.c_str(), 1, 1);
+   mesh->PrintInfo();
+
+   // Print mesh to file for visualization
+   if (visit)
+   {
+      VisItDataCollection dc = VisItDataCollection("mesh", mesh);
+      dc.SetPrefixPath("Naca_cmesh");
+      dc.SetCycle(0);
+      dc.SetTime(0.0);
+      dc.Save();
+   }
+
+   if (visualization)
+   {
+      // Create glvis output
+      string glvis_file;
+      glvis_file.append(msh_path);
+      glvis_file.append("glvis_");
+      glvis_file.append(msh_filename);
+      glvis_file.append(".mesh");
+      ofstream glvis_output(glvis_file.c_str());
+      mesh->Print(glvis_output);
+      glvis_output.close();
+   }
+
+   delete mesh;
+   return 0;
+}
+
+NACA4::NACA4(double t_, double c_)
+{
+   t = t_;
+   c = c_;
+   A = 0.2969, B = 0.1260, C = 0.3516, D = 0.2843, E = 0.1036;
+   iter_max = 1000;
+   epsilon = 1e-8;
+}
+
+double NACA4::y(double x) const
+{
+   double y = 5*t*(A*sqrt(x/c) - B*x/c - C*pow(x/c,2)
+                   + D*pow(x/c,3) - E*pow(x/c,4));
+   return y*c;
+}
+
+double NACA4::dydx(double x) const
+{
+   double y = 5*t*(0.5 * A/sqrt(x/c) - B - 2*C*x/c
+                   + 3* D*pow(x/c,2) - 4* E*pow(x/c,3));
+   return y*c;
+}
+
+double NACA4::len(double x) const
+{
+   double l = 5 * t * (A*sqrt(x/c) - B*x - C*pow(x/c,2)
+                       + D*pow(x/c,3) - E * pow(x/c,4)) + x/c;
+   return l*c;
+}
+
+double NACA4::dlendx(double xi) const
+{
+   return 1 + dydx(xi);
+}
+
+double NACA4::xl(double l) const
+{
+   double x = l; // Initial guess, length should be a good one
+   double h;
+   int i = 0;
+   do
+   {
+      x = abs(x); // The function and its derivative do not exist for x < 0
+      // Newton step: x(i+1) = x(i) - f(x) / f'(x)
+      h = (len(x) - l)/dlendx(x);
+      x = x - h;
+   }
+   while (abs(h) >= epsilon && i++ < iter_max);
+
+   if (i >= iter_max) { mfem_error("Did not find root"); }
+   return x;
+}
+
+void GetTipXY(const NACA4 &foil_section, KnotVector *kv, double tf,
+              Array<Vector*> &xy)
+{
+   int ncp = kv->GetNCP();
+   // Length of half the curve: the boundary covers both sides of the tip
+   const double  llll = foil_section.GetChord();
```

Comment:
Unused.

Reply:
This was still work in progress after I was summoned for diner ;)